src/nodes/dir.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License.  You may obtain a copy
+// of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+extern crate fuse;
+extern crate libc;
+extern crate time;
+
+use std::ffi::OsStr;
+use std::io;
+use std::sync::{Arc, Mutex};
+use self::time::Timespec;
+use super::Node;
+
+/// Representation of a directory node.
+pub struct Dir {
+    inode: u64,
+    state: Mutex<MutableDir>,
+}
+
+/// Holds the mutable data of a directory node.
+struct MutableDir {
+    parent: u64,
+    attr: fuse::FileAttr,
+}
+
+impl Dir {
+    /// Creates a new scaffold directory identified by the given `inode`.
+    ///
+    /// `parent` specifies the inode number of the parent directory which may be the same as
+    /// `inode` when this directory represents the root of the file system.
+    ///
+    /// `time` is the timestamp to be used for all node times.
+    ///
+    /// `uid` and `gid` indicate the ownership details of the node.  For scaffold directories, these
+    /// should always match the values of the currently-running process -- but not necessarily if we
+    /// want to let users customize these via flags at some point.
+    ///
+    /// Scaffold directories are in-memory directories not backed by any on-disk contents.  These
+    /// are used to represent user-specified mappings
+    pub fn new_empty(inode: u64, parent: u64, time: Timespec, uid: u32, gid: u32) -> Arc<Node> {
```

Comment:
Maybe this should be new_root? I feel like it would be safer to have a "mkdir" method to make child directories.

Reply:
This won't be used exclusively for the root directory though.

There are two variants of directory nodes: "mapped" directories, which correspond to a "real" directory in the underlying file system and which might have explicit overrides for certain entries; and "scaffold" directories, which are pure in-memory directories that initially start empty and gain entries (mapped or scaffold) as the user applies new mappings to them.

These two concepts were originally separate in the Go implementation as two different node types, but that caused a lot of problems to represent the root directory: during a "reconfiguration" (when we change the mappings the file system exposes without a remount), the type of the directories can change -- and the kernel (or rather, the FUSE library) doesn't like it if the type of the root directory changes. Hence why the two abstractions were merged. See https://github.com/jmmv/sandboxfs/commit/5b9f460b92853e814f1979786b5adf55ce3da912, https://github.com/jmmv/sandboxfs/commit/4d5042e21bea2d5c5808eb898b819e287c06dd5d, and https://github.com/jmmv/sandboxfs/commit/f9d56c7da24db26351eea3e6a1283c364519a9a5 for some context.

Actually, merging the two types brought in a lot of simplicity to the code: both directory types behave pretty much in the same manner and the only difference is whether one contains "real" entries or not. The restriction that the "root cannot change" doesn't apply in this Rust implementation because we are defining our own nodes and can do as we please, but I think it makes sense to keep them as a single entity for simplicity.

Thoughts for a better name welcome, but I think the context can help here.

Reply:
I'm not suggesting separate types, just separate methods. `parent: u64` is just asking for a referential integrity violation. Having "it's a root directory or you created it from another node" makes more sense. I suppose you could use `parent: Option<Node>` or similar, but it makes more sense to me to have separate `make_root` and `Node.mkdir` functions.

Reply:
Oh I see, good point. OK, changed to `new_root` and simplified the API for now. Once the ability to create subdirectories/map entries exists, I'll figure out how to refactor this code to reuse whatever is possible.